File: LinkedList.cpp
#include "LinkedList.hpp"

#include <stdexcept>

Node::Node() : value_(0), next_(nullptr) {
}

Node::Node(const int value) : value_(value), next_(nullptr) {
}

Node::~Node() {
    value_ = 0;
    next_ = nullptr;
}

void Node::SetValue(const int value) {
    value_ = value;
}

void Node::SetNext(Node* node) {
    next_ = node;
}

int Node::GetValue() {
    return value_;
}

Node* Node::GetNext() {
    return next_;
}

LinkedList::LinkedList() : begin_(nullptr), end_(nullptr) {
}

LinkedList::LinkedList(Node* node) : begin_(node), end_(node) {
}

LinkedList::LinkedList(const LinkedList& other) : begin_(nullptr), end_(nullptr) {
    if (other.begin_ == nullptr) {
        throw std::runtime_error("Invalid linked list for constructor");
    }

    Node* current = other.begin_;
    while (current != nullptr) {
        this->PushBack(current->GetValue());
        current = current->GetNext();
    }
}

LinkedList& LinkedList::operator=(const LinkedList& other) {
    if (this == &other) {
        return *this;
    }

    this->~LinkedList();

    Node* current = other.begin_;
    while (current != nullptr) {
        this->PushBack(current->GetValue());
        current = current->GetNext();
    }

    return *this;
}

LinkedList::LinkedList(LinkedList&& other) noexcept : begin_(other.GetBegin()), end_(other.GetEnd()) {
    other.begin_ = nullptr;
    other.end_ = nullptr;
}

LinkedList& LinkedList::operator=(LinkedList&& other) noexcept {
    if (begin_ != nullptr) {
        this->~LinkedList();
    }

    if (this == &other) {
        return *this;
    }

    begin_ = other.GetBegin();
    end_ = other.GetEnd();

    other.begin_ = nullptr;
    other.end_ = nullptr;

    return *this;
}

LinkedList::~LinkedList() {
    Node* current = begin_;

    while (current != nullptr) {
        Node* temp = current;
        current = current->GetNext();

        delete temp;
    }
}

Node* LinkedList::GetBegin() {
    return begin_;
}

Node* LinkedList::GetEnd() {
    return end_;
}

void LinkedList::PushBack(const int value) {
    if (begin_ == nullptr) {
        Node* node = new Node(value);
        begin_ = node;
        end_ = node;

        return;
    }

    Node* node = new Node(value);
    end_->SetNext(node);
    end_ = node;
}

void LinkedList::Insert(const int value, const size_t position) {
    if (begin_ == nullptr && position > 0) {
        throw std::runtime_error("Invalid position for insertion");
    }

    Node* node = new Node(value);

    if (begin_ == nullptr) {
        begin_ = node;

        if (end_ == nullptr) {
            end_ = node;
        }

        return;
    }

    if (position == 0) {
        node->SetNext(begin_);
        begin_ = node;

        if (end_ == nullptr) {
            end_ = node;
        }

        return;
    }

    Node* current = begin_;
    for (size_t i = 0; i < position - 1; ++i) {
        current = current->GetNext();

        if (current == nullptr) {
            throw std::runtime_error("Invalid position for insertion");
        }
    }

    node->SetNext(current->GetNext());
    current->SetNext(node);
}

void LinkedList::Replace(const int value, const size_t position) {
    if (begin_ == nullptr) {
        throw std::runtime_error("Invalid linked list for replacement");
    }

    if (position == 0) {
        begin_->SetValue(value);

        return;
    }

    Node* current = begin_;
    for (size_t i = 0; i < position; ++i) {
        current = current->GetNext();

        if (current == nullptr) {
            throw std::runtime_error("Invalid position for replacement");
        }
    }

    current->SetValue(value);
}

void LinkedList::Delete(const size_t position) {
    if (begin_ == nullptr) {
        throw std::runtime_error("Invalid linked list for deleting an element");
    }

    if (position == 0) {
        Node* temp = begin_;
        begin_ = begin_->GetNext();

        delete temp;
        temp = nullptr;

        return;
    }

    Node* current = begin_;
    for (size_t i = 0; i < position - 1; ++i) {
        current = current->GetNext();

        if (current == nullptr) {
            throw std::runtime_error("Invalid position for deleting an element");
        }
    }

    if (current->GetNext()->GetNext() == nullptr) {
        delete current->GetNext();
        current->SetNext(nullptr);

        end_ = current;

        return;
    }

    Node* temp = current->GetNext();

    current->SetNext(current->GetNext()->GetNext());

    delete temp;
    temp = nullptr;
}

std::ostream& operator<<(std::ostream& os, const LinkedList& list) {
    Node* current = list.begin_;

    while (current != nullptr) {
        os << current->GetValue() << ' ';
        current = current->GetNext();
    }

    return os;
}

File: go/LinkedList/LinkedList.go
package LinkedList

import (
	"errors"
	"fmt"
)

type Node[T any] struct {
	value T
	next  *Node[T]
}

type LinkedList[T any] struct {
	begin *Node[T]
	end   *Node[T]
}

func New[T any]() *LinkedList[T] {
	return &LinkedList[T]{begin: nil, end: nil}
}

func (list *LinkedList[T]) GetBegin() *Node[T] {
	return list.begin
}

func (list *LinkedList[T]) GetEnd() *Node[T] {
	return list.end
}

func (list *LinkedList[T]) Find(value T, equals func(a, b T) bool) (int, error) {
	if list.begin == nil {
		return -1, errors.New("invalid list for search")
	}

	current := list.begin
	for index := 0; current != nil; index++ {
		if equals(current.value, value) {
			return index, nil
		}
		current = current.next
	}

	return -1, errors.New("element not found")
}

func (list *LinkedList[T]) PushBack(value T) {

	node := &Node[T]{value: value, next: nil} // new node

	if list.begin == nil {
		list.begin = node
		list.end = node

		return
	}

	list.end.next = node
	list.end = node
}

func (list *LinkedList[T]) Insert(value T, position int) error {
	if list.begin == nil && position != 0 || position < 0 {
		return errors.New("invalid position for insertion")
	}

	node := &Node[T]{value: value, next: nil} // new node

	if list.begin == nil {
		list.begin = node
		list.end = node

		return nil
	}

	if position == 0 {
		node.next = list.begin
		list.begin = node

		return nil
	}

	current := list.begin
	for index := 0; current != nil && index < position-1; index++ {
		current = current.next
	}

	if current == nil {
		return errors.New("invalid position for insertion")
	}

	node.next = current.next
	current.next = node

	return nil
}

func (list *LinkedList[T]) Replace(value T, position int) error {
	if list.begin == nil || position < 0 {
		return errors.New("invalid position for insertion")
	}

	current := list.begin
	for index := 0; current != nil && index < position; index++ {
		current = current.next
	}

	if current == nil {
		return errors.New("invalid position for insertion")
	}

	current.value = value

	return nil
}

func (list *LinkedList[T]) Delete(position int) error {
	if list.begin == nil || position < 0 {
		return errors.New("invalid position for deletion")
	}

	current := list.begin

	if position == 0 {
		list.begin = current.next
		if list.begin == nil {
			list.end = nil
		}

		return nil
	}

	for index := 0; current != nil && index < position-1; index++ {
		current = current.next
	}

	if current == nil {
		return errors.New("invalid position for deletion")
	}

	if current.next.next == nil {
		current.next = nil
		list.end = current

		return nil
	}

	current.next = current.next.next

	return nil
}

func (list *LinkedList[T]) String() string {
	if list.begin == nil {
		return "[]"
	}

	result := "["

	current := list.begin
	for current != nil {
		result += fmt.Sprintf("%v", current.value)

		if current.next != nil {
			result += ", "
		}

		current = current.next
	}

	return result + "]"
}
